File: src/main/java/itmo/is/project/service/user/UserService.java
package itmo.is.project.service.user;

import itmo.is.project.dto.security.RegistrationRequest;
import itmo.is.project.mapper.user.UserMapper;
import itmo.is.project.model.user.Role;
import itmo.is.project.model.user.User;
import itmo.is.project.repository.user.UserRepository;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;
    private final UserMapper userMapper;
    private final PasswordEncoder passwordEncoder;
    private final AccountService accountService;

    public User findUserById(Integer userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new EntityNotFoundException("User not found with id: " + userId));
    }

    public User findUserByUsername(String username) {
        return userRepository.findByUsername(username)
                .orElseThrow(() -> new EntityNotFoundException("User not found with username: " + username));
    }

    public Page<User> findAllDisabledUsers(Pageable pageable) {
        return userRepository.findAllByEnabledFalse(pageable);
    }

    public boolean isOwnerRegistered() {
        return userRepository.existsByRole(Role.ROLE_OWNER);
    }

    public void updateUser(User user) {
        userRepository.save(user);
    }

    public void deleteUser(User user) {
        userRepository.delete(user);
    }

    public User createUser(RegistrationRequest request, Role role, boolean enabled) {
        validateUsername(request.username());
        User user = userRepository.save(toUser(request, role, enabled));
        createAccountIfPilotOrOwner(user);
        return user;
    }

    private void validateUsername(String username) {
        if (userRepository.existsByUsername(username)) {
            throw new IllegalArgumentException("Username is taken: " + username);
        }
    }

    private User toUser(RegistrationRequest request, Role role, boolean enabled) {
        User user = userMapper.toEntity(request);
        user.setPassword(passwordEncoder.encode(user.getPassword()));
        user.setRole(role);
        user.setEnabled(enabled);
        return user;
    }

    private void createAccountIfPilotOrOwner(User user) {
        if (user.getRole() == Role.ROLE_PILOT || user.getRole() == Role.ROLE_OWNER) {
            accountService.createAccount(user);
        }
    }
}

File: src/main/kotlin/itmo/isproject/service/user/UserService.kt
package itmo.isproject.service.user

import itmo.isproject.dto.security.RegistrationRequest
import itmo.isproject.dto.user.UserDto
import itmo.isproject.mapper.user.UserMapper
import itmo.isproject.model.user.Role
import itmo.isproject.model.user.User
import itmo.isproject.repository.user.UserRepository
import jakarta.persistence.EntityNotFoundException
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class UserService(
    private val userRepository: UserRepository,
    private val userMapper: UserMapper,
    private val passwordEncoder: PasswordEncoder,
    private val accountService: AccountService
) {

    fun getAllUsers(pageable: Pageable, role: Role?): Page<UserDto> {
        return if (role == null) {
            userRepository.findAll(pageable).map { userMapper.toDto(it) }
        } else {
            userRepository.findAllByRole(role, pageable).map { userMapper.toDto(it) }
        }
    }

    fun getUserById(id: Int): UserDto {
        return userMapper.toDto(findUserById(id))
    }

    fun findUserById(userId: Int): User {
        return userRepository.findByIdOrNull(userId)
            ?: throw EntityNotFoundException("User not found with id: $userId")
    }

    fun findUserByUsername(username: String): User {
        return userRepository.findByUsernameInternal(username)
            ?: throw EntityNotFoundException("User not found with username: $username")
    }

    fun findAllDisabledUsers(pageable: Pageable): Page<User> {
        return userRepository.findAllByEnabledFalse(pageable)
    }

    fun isOwnerRegistered(): Boolean {
        return userRepository.existsByRole(Role.ROLE_OWNER)
    }

    fun updateUser(user: User) {
        userRepository.save(user)
    }

    fun deleteUser(user: User) {
        userRepository.delete(user)
    }

    fun createUser(request: RegistrationRequest, role: Role, enabled: Boolean): User {
        validateUsername(request.username)
        val user = userRepository.save(toUser(request, role, enabled))
        createAccountIfPilotOrOwner(user)
        return user
    }

    private fun validateUsername(username: String?) {
        if (userRepository.existsByUsernameInternal(username)) {
            throw IllegalArgumentException("Username is taken: $username")
        }
    }

    private fun toUser(request: RegistrationRequest, role: Role, enabled: Boolean): User {
        val user = userMapper.toEntity(request)
        user.passwordInternal = passwordEncoder.encode(user.passwordInternal)
        user.role = role
        user.enabled = enabled
        return user
    }

    private fun createAccountIfPilotOrOwner(user: User) {
        if (user.role == Role.ROLE_PILOT || user.role == Role.ROLE_OWNER) {
            accountService.createAccount(user)
        }
    }
}
